src/bgw/scheduler.c

```diff
@@ -247,6 +249,7 @@ scheduled_bgw_job_transition_state_to(ScheduledBgwJob *sjob, JobState new_state)
 				elog(WARNING,
 					 "scheduler detected that job %d was deleted when starting job",
 					 sjob->job.fd.id);
+				ts_bgw_job_cache_invalidate_callback();
```

Comment:
Why aren't we using `ts_catalog_invalidate_cache()` so that this invalidation is propagated to all backends and processes? Also, it seems to me that this shouldn't be necessary since we already have a callback to invalidate and call this function whenever a DELETE is issued on the `bgw_job` table.

Reply:
@erminator This invalidates the bgw job cache since we detected a potential missing job. (This ideally should not be necessary since we use the ts_catalog_invalidate_cache callback for deletes as you said. But we appear to be missing some cache invalidations and try to schedule jobs that have been deleted in some cases. This code is trying to be defensive and refresh the job cache when we detect a potential problem).

Reply:
I am worried that we are just brushing the real problem under the rug by being defensive. 

I don't get the "potential missing job"... Either it is missing or it is not. There must be a source of truth and I don't understand why we are trying to reschedule it here if it is indeed missing. The fact that the cache was stale just seems to be a follow-on problem to not doing the right thing here.

`ts_bgw_job_get_share_lock` seems to scan the jobs table so either that code is broken and returns the wrong information. I guess that code actually conflates two things:

- Unable to grab lock
- Job missing

I suggest returning a status enum from `ts_bgw_job_get_share_lock` that tells us which case we are dealing with and only reschedule in case the job existed but we couldn't grab the lock.

Reply:
I agree that rescheduling was not the right thing. Gayathri removed that. This code is blocking so `unable to grab lock` is not a possibility. Gayathri fixed what I think was the cache problem in a previous PR. I think it's still worth being defensive here as this is a huge problem for customers on cloud if it does occur as there is no way to restart the server to recover.

Reply:
would that not happen when you have lock_timeout set to a non-zero value

Reply:
I believe that that timeout throws an error if lock timeout is reached